src/mlpack/core/optimizers/smorms3/smorms3_impl.hpp

```diff
@@ -0,0 +1,144 @@
+    function(function),
+    lRate(lRate),
+    eps(eps),
+    maxIterations(maxIterations),
+    tolerance(tolerance),
+    shuffle(shuffle)
+{ /* Nothing to do. */ }
+
+//! Optimize the function (minimize).
+template<typename DecomposableFunctionType>
+double SMORMS3<DecomposableFunctionType>::Optimize(arma::mat& iterate)
+{
+  // Find the number of functions to use.
+  const size_t numFunctions = function.NumFunctions();
+
+  // This is used only if shuffle is true.
+  arma::Col<size_t> visitationOrder;
+  if (shuffle)
+    visitationOrder = arma::shuffle(arma::linspace<arma::Col<size_t>>(0,
+        (numFunctions - 1), numFunctions));
+
+  // To keep track of where we are and how things are going.
+  size_t currentFunction = 0;
+  double overallObjective = 0;
+  double lastObjective = DBL_MAX;
+
+  // Calculate the first objective function.
+  for (size_t i = 0; i < numFunctions; ++i)
+    overallObjective += function.Evaluate(iterate, i);
+
+  // Initialise the parameters mem, g and g2.
+  arma::mat mem = arma::ones<arma::mat>(iterate.n_rows, iterate.n_cols);
+  arma::mat g = arma::zeros<arma::mat>(iterate.n_rows, iterate.n_cols);
+  arma::mat g2 = arma::zeros<arma::mat>(iterate.n_rows, iterate.n_cols);
+
+  // Now iterate!
+  arma::mat gradient(iterate.n_rows, iterate.n_cols);
+  
+  for (size_t i = 1; i != maxIterations; ++i, ++currentFunction)
+  {
+    // Is this iteration the start of a sequence?
+    if ((currentFunction % numFunctions) == 0)
+    {
+      // Output current objective function.
+      Log::Info << "SMORMS3: iteration " << i << ", objective "
+          << overallObjective << "." << std::endl;
+
+      if (std::isnan(overallObjective) || std::isinf(overallObjective))
+      {
+        Log::Warn << "SMORMS3: converged to " << overallObjective
+            << "; terminating with failure. Try a smaller step size?"
+            << std::endl;
+        return overallObjective;
+      }
+
+      if (std::abs(lastObjective - overallObjective) < tolerance)
+      {
+        Log::Info << "SMORMS3: minimized within tolerance " << tolerance << "; "
+            << "terminating optimization." << std::endl;
+        return overallObjective;
+      }
+
+      // Reset the counter variables.
+      lastObjective = overallObjective;
+      overallObjective = 0;
+      currentFunction = 0;
+
+      if (shuffle) // Determine order of visitation.
+        visitationOrder = arma::shuffle(visitationOrder);
+    }
+
+    // Evaluate the gradient for this iteration.
+    if (shuffle)
+      function.Gradient(iterate, visitationOrder[currentFunction], gradient);
+    else
+      function.Gradient(iterate, currentFunction, gradient);
+
+    // And update the iterate.
+    arma::mat r = 1 / (mem + 1);
+
+    g = (1 - r) % g;
+    g += r % gradient;
+
+    g2 = (1 - r) % g2;
+    g2 += r % (gradient % gradient);
+
+    arma::mat x = (g % g) / (g2 + eps);
+
+    arma::mat lRateMat(x.n_rows, x.n_cols);
```

Comment:
> It's odd that we have to create another matrix of size x

I think we could do with size of iterate as well, as both are really the same size matrices (x and iterate).
> I think we should avoid the recreation of the learning matrix for every iteration, instead we could create it once and update it every iteration. What do you think?

You are right. In fact, I see that we don't need to update or fill `lRateMat` with `lRate` in every iteration as `lRate` doesn't change during the iterations. So, both the statements in line 116 and 117 can be moved outside the for loop. What do you suggest?

P.S. Sorry for the delayed response. Have been busy with preparing for exams starting from tomorrow.